### PR TITLE
Refactor LinuxTracingHandler, LinuxTracingSession, and co.

### PR DESCRIPTION
**Pull request by dpallotti**
- Cut dependencies of `LinuxTracingHandler` on `SamplingProfiler` and `Process`.
  These classes store information in the client and in Windows profiling, but on
  the service all the profiling information is held by `LinuxTracingHandler` and
  `LinuxTracingSession`.
- Use `LinuxTracingHandler` directly in the `ConnectionManager` for now (this
  should probably be changed next) instead of going through `EventTracer`. This
  had a role on Windows profiling and is still used in the UI, but is not needed
  by the service.
- Similarly, do not depend on `Capture` in `ConnectionManager` to start and stop
  a trace. Use `LinuxTracingHandler` directly instead.
- Align the way `LinuxAddressInfo`s and `KeyAndString`s are sent to the way
  the other events are sent, i.e. buffered in `LinuxTracingSession` and sent in
  batches by the `ConnectionManager`. This fixes some TODOs.
- `join()` `OrbitLinuxTracing/Tracer::thread_`.
- Refactor `StringManager`. In particular, change `Add` to `AddIfNotContains`.
- Refactor `LinuxCallstackEvent`: remove unused fields.
- Rename `LinuxTracingSession` to `LinuxTracingBuffer` to reflect its usage.
- Make `TransactionManager`'s deleted methods public.
- `clang-tidy`, `clang-format`, and small code tweaks here and there.

### CODE REVIEW

**Comment by dimitry-** [comment]
(it does not help that github does not detect file renames)

**Comment by dpallotti** [comment]
> Hi, looks good but very hard to review given that it has so many changes at once.
> Can you please split this change to more moderately-sized commits?

I'm aware and I'm sorry. Unfortunately for these big refactorings where a lot of changes depend on each other and are done in parallel, trying to keep separate commits and rebasing fixup commits becomes incredibly time consuming and frustrating, as with more than a couple of commits one starts continuously running into rebase conflicts. You are right that some could have easily been separate commits, above all the renaming of `LinuxTracingSession`, but bear with me.

File: OrbitCore/ConnectionManager.cpp [hunk 1]
```diff
@@ -97,72 +94,79 @@ void ConnectionManager::SetSelectedFunctionsOnRemote(const Message& a_Msg) {
...
+      GTcpServer->Send(Msg_KeysAndStrings, message_data.c_str(),
+                       message_data.size());
+    }
   }
+#endif
 }
 
 void ConnectionManager::SetupIntrospection() {
-#if __linux__ && ORBIT_TRACING_ENABLED
+#if defined(__linux__) && ORBIT_TRACING_ENABLED
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/ConnectionManager.cpp:145]: In OrbitBase/include/OrbitBase/Tracing.h we have `#define ORBIT_TRACING_ENABLED 0`, that's why `defined` is not used. Should I remove `#define ORBIT_TRACING_ENABLED 0` and switch all usages to `ifdef`s?

File: OrbitCore/LinuxTracingHandler.cpp [hunk 2]
```diff
@@ -112,29 +94,35 @@ void LinuxTracingHandler::OnCallstack(
...
-                       message_data.size());
-
-      target_process_->AddAddressInfo(std::move(address_info));
+    {
+      std::lock_guard<std::mutex> lock(addresses_seen_mutex_);
+      if (!addresses_seen_.contains(address)) {
+        LinuxAddressInfo address_info{address, frame.GetMapName(),
+                                      llvm::demangle(frame.GetFunctionName()),
+                                      frame.GetFunctionOffset()};
+        tracing_buffer_->RecordAddressInfo(std::move(address_info));
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/LinuxTracingHandler.cpp:105]: The reason of this is that I think it's more readable, the same below with `hashed_cs_event` and `cs_event`. What do you think?

File: OrbitCore/ConnectionManager.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dimitry-) [reply OrbitCore/ConnectionManager.cpp:145]: something for the another PR then.

File: OrbitCore/LinuxTracingHandler.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dimitry-) [reply OrbitCore/LinuxTracingHandler.cpp:105]: I feel like the method name provides enough context - but this is up to you. maybe preserve it for uniformity, the below code has a bit different naming for classes and methods.

File: OrbitCore/LinuxTracingHandler.h [hunk 3]
```diff
@@ -44,20 +39,20 @@ class LinuxTracingHandler : LinuxTracing::TracerListener {
...
-  Process* target_process_;
-  std::map<uint64_t, Function*>* selected_function_map_;
-  uint64_t* num_context_switches_;
+  uint64_t ProcessStringAndGetKey(const std::string& string);
 
+  LinuxTracingBuffer* tracing_buffer_;
   std::unique_ptr<LinuxTracing::Tracer> tracer_;
 
+  absl::flat_hash_set<uint64_t> addresses_seen_;
+  std::mutex addresses_seen_mutex_;
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/LinuxTracingHandler.h]: Done, also for StringManager.

File: OrbitCore/LinuxTracingHandler.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/LinuxTracingHandler.cpp:106]: Then I'll leave it as it is for consistency.

File: OrbitCore/ConnectionManager.cpp [hunk 4]
```diff
@@ -205,53 +209,37 @@ void ConnectionManager::SetupClientCallbacks() {
...
+    }
   });
 
-  GTcpClient->AddCallback(Msg_RemoteCallStack, [=](const Message& a_Msg) {
-    CallStack stack;
+  GTcpClient->AddCallback(Msg_LinuxAddressInfos, [=](const Message& a_Msg) {
     std::istringstream buffer(a_Msg.GetDataAsString());
-    cereal::JSONInputArchive inputAr(buffer);
-    inputAr(stack);
+    cereal::BinaryInputArchive inputAr(buffer);
```
Comment (pierricgimmig) [inline OrbitCore/ConnectionManager.cpp:233]: You could also use DeserializeObjectBinary() from Serialization.h, but it takes in a void* and a size, we could add a version that takes in a string.

File: OrbitCore/EventBuffer.cpp [hunk 5]
```diff
@@ -85,37 +85,13 @@ ORBIT_SERIALIZE(EventBuffer, 0) {
 }
 
 //-----------------------------------------------------------------------------
-ORBIT_SERIALIZE(CallstackEvent, 0) {
-  ORBIT_NVP_VAL(0, m_Time);
+ORBIT_SERIALIZE(CallstackEvent, 1) {
```
Comment (pierricgimmig) [inline OrbitCore/EventBuffer.cpp:88]: Seems like there is a version bump but no actual change in Serialization.

File: OrbitCore/StringManager.cpp [hunk 6]
```diff
@@ -1,14 +1,16 @@
 #include "StringManager.h"
 
-void StringManager::Add(uint64_t key, std::string_view str) {
-  std::lock_guard<std::mutex> lock(mutex_);
-  if (key_to_string_.count(key) == 0) {
-    key_to_string_.emplace(key, str);
+bool StringManager::AddIfNotContains(uint64_t key, std::string_view str) {
```
Comment (pierricgimmig) [inline OrbitCore/StringManager.cpp:3]: AddIfNotPresent?

File: OrbitCore/ConnectionManager.cpp (hunk 4 above)
Comment (dpallotti) [reply OrbitCore/ConnectionManager.cpp]: Added `void DeserializeObjectBinary(const std::string& data, T& object)` and used in these four `GTcpClient->AddCallback`.

File: OrbitCore/EventBuffer.cpp (hunk 5 above)
Comment (pierricgimmig) [reply OrbitCore/EventBuffer.cpp:88]: Oh sorry, I missed that.  All good!

**Pull request merged**